### ISSUE TITLE
[Good First Issue] Cover NNCF code with least coverage percentage with unit tests to at least 80%

### ISSUE THREAD

**Issue by vshampor** [openvinotoolkit/openvino#22194]
## Context

[NNCF](https://github.com/openvinotoolkit/nncf) uses `coverage` to run its precommit checks and currently has an admirable coverage rate of 90% across the repository. However, there are still some files that have <80% coverage, some - with no coverage at all. Some of these are located in the `experimental` folder and do not have a strict coverage requirement, some are special files such as `__init__.py` or `setup.py` which are not easily unit-tested, but most have actual production code inside and many of those have exactly 0.00% coverage. We need such files covered by precommit-scope unit tests in order not to let potential bugs get away.

## Task
1. Review the current per-file coverage percentage at Codecov (sort by coverage)
https://app.codecov.io/gh/openvinotoolkit/nncf?search=&displayType=list

2. For the following files in the list (excluding experimental and special files), add unit tests (using `pytest` style) under the `tests` directory (in a subdirectory corresponding to the tested backend, e.g. a test for `nncf/torch/foo.py` would go to `tests/torch/test_foo.py`, for `nncf/common/bar.py` - to `tests/torch/test_bar.py`).

- [ ] [nncf/common/utils/decorators.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/decorators.py)
- [ ] [nncf/common/utils/logger/__init__.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/logger/__init__.py)
- [ ] [nncf/common/utils/os.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/os.py)
- [ ] [nncf/common/utils/patcher.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/patcher.py)
- [ ] [nncf/common/utils/timer.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/timer.py)
- [ ] [nncf/config/schema.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/config/schema.py)
- [ ] [nncf/openvino/pot/quantization/accuracy_aware.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/openvino/pot/quantization/accuracy_aware.py)
- [ ] [nncf/openvino/pot/quantization/quantize_model.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/openvino/pot/quantization/quantize_model.py)
- [ ] [nncf/openvino/quantization/quantize_model.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/openvino/quantization/quantize_model.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/algorithm.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/evaluator.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/evaluator.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/openvino_backend.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/openvino_backend.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/ranker.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/ranker.py)
- [ ] [nncf/quantization/algorithms/accuracy_control/subset_selection.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/subset_selection.py)
- [ ] [nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py)
- [ ] [nncf/quantization/algorithms/hyperparameter_tuner/param_grid.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/hyperparameter_tuner/param_grid.py)
- [ ] [nncf/quantization/algorithms/post_training/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/post_training/algorithm.py)
- [ ] [nncf/telemetry/wrapper.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/telemetry/wrapper.py)
- [ ] [nncf/tensorflow/__init__.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/__init__.py)
- [ ] [nncf/tensorflow/helpers/model_creation.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/helpers/model_creation.py)
- [ ] [nncf/tensorflow/pruning/filter_pruning/functions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/pruning/filter_pruning/functions.py)
- [ ] [nncf/tensorflow/tensor_statistics/statistics.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/tensor_statistics/statistics.py)
- [ ] [nncf/tensorflow/tf_internals.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/tf_internals.py)
- [ ] [nncf/tensorflow/utils/node.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/utils/node.py)
- [ ] [nncf/tensorflow/utils/scopes_handle.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/tensorflow/utils/scopes_handle.py)
- [ ] [nncf/torch/accuracy_aware_training/utils.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/accuracy_aware_training/utils.py)
- [ ] [nncf/torch/binarization/binarize_functions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/binarization/binarize_functions.py)
- [ ] [nncf/torch/binarization/extensions.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/binarization/extensions.py)
- [ ] [nncf/torch/pruning/export_utils.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/pruning/export_utils.py)
- [ ] [nncf/torch/pruning/operations.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/pruning/operations.py)
- [ ] [nncf/torch/quantization/base_ctrl.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/base_ctrl.py)
- [ ] [nncf/torch/quantization/precision_init/autoq_init.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/precision_init/autoq_init.py)
- [ ] [nncf/torch/quantization/statistics.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/quantization/statistics.py)
- [ ] [nncf/torch/sparsity/const/algo.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/const/algo.py)
- [ ] [nncf/torch/sparsity/rb/algo.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/rb/algo.py)
- [ ] [nncf/torch/sparsity/rb/loss.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/rb/loss.py)

Start with the files having the least amount of coverage. 

Make sure that the test cases you are adding actually execute the lines that are marked as "missed" in the Codecov per-line display (for instance, [nncf/quantization/algorithms/accuracy_control/ranker.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf%2Fquantization%2Falgorithms%2Faccuracy_control%2Franker.py)). 

If at all possible, try to extend existing parametrized tests with proper parameters triggering the missed line logic. 

Try to add new test cases to the existing test files that contain the tests for the tested component, otherwise - create a new test file in a proper subdirectory with a proper name, use existing test files as a guideline.

If you encounter dead code (for example, helper functions that aren't used anywhere, only defined) - delete it instead of covering it with tests. This will raise the coverage for the corresponding file better than anything.

**Comment by Saharsh365** [comment]
I'd like to be assigned this issue. 
Also, since this will be my first time contributing, do you have any advice for me? And, is there a discord or other means for communication for the community I can join?

**Comment by Saharsh365** [comment]
.take

**Comment by mlukasze** [comment]
> And, is there a discord or other means for communication for the community I can join?

yes sir! [Discord](https://discord.gg/7pVRxUwdWG)

**Comment by MaximProshin** [comment]
@Saharsh365 , thanks for your interest! The description is quit detailed and there are also tips. My recommendation would be to start from the analysis of the current codecov results, selecting one of the weak files and implementing the first simple test. If it works, the rest could be done by analogy. @vshampor will support you in case of questions. Don't hesitate to bring your questions.

**Comment by Saharsh365** [comment]
@vshampor The code coverage analysis reports are not working for me anymore. Could you please help me solve this issue

**Comment by vshampor** [comment]
@Saharsh365 we won't be able to help you unless you describe the problem you are encountering in a more specific fashion. I checked that the reports are visible for non-authenticated viewers.

**Comment by Saharsh365** [comment]
> @Saharsh365 we won't be able to help you unless you describe the problem you are encountering in a more specific fashion. I checked that the reports are visible for non-authenticated viewers.

Oh, I'm not sure what happened but I believe that from Friday evening to yesterday the reports were not being displayed. I can however see them now and I believe that may have been due to the weekend? Apologies for taking up your time.

**Comment by Saharsh365** [comment]
I would like to drop this issue since I am not sure about how to proceed with this task. I used multiple resources to research how to write code coverage tests and I am still unsure about how to apply my knowledge to this codebase to write coverage tests

**Comment by p-wysocki** [comment]
Hello @Saharsh365, I'm unassigning you, but feel free to repick the task if you wish to - we're here to answer questions, perhaps asking them on Intel DevHub channel would be easier.

**Comment by Candyzorua** [comment]
.take

**Comment by Candyzorua** [comment]
Hello! I am a first-time contributor and I'm excited to help! 

I have written unit tests for [nncf/common/utils/decorators.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/decorators.py) and [nncf/common/utils/os.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/os.py). 

May I know if there are other ways to check the code coverage before making the PR, other than coverage.py? Would it be advised for me to open a PR for just these 2 files, to check if I am on the right track? :)

I'm also partway through making test for [nncf/common/utils/patcher.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/common/utils/patcher.py) and I'd like to ask some questions on it. Particularly, I'm not clear regarding the `patch` method.

For the `patch` method, it seems that there are 3 separate cases: module, class, and instance-level functions. However, I'm not sure when the instance-level case should be triggered, as when an instance method is passed as the 2nd argument to patch, the class-level case is still triggered due to line 44, `obj_cls, fn_name = self.import_obj(obj_cls)` making `obj_cls` into a Class object and line 55 always evaluating to True.

Could this method possibly be bugged or am I understanding it incorrectly? Thank you for reading, really appreciate any help.

**Comment by vshampor** [comment]
Greetings!

The implementation may have subtle bugs, in particular because the coverage is missing. I suggest that you add a test case for it. Here's how the "instance method" patching supposed to start from:
```python3
class Model:
    def forward(self, x):
        return x

model = Model()
assert model.forward(7) == 7  # `forward` is looked up on the Model.__dict__, i.e. the class object

# now someone (not us) patches the forward with its own implementation:
def new_forward(x):
    return 0

model.forward = new_forward  # creates a new *instance* attribute `forward` which masks the original function definition from the class

assert model.forward(7) == 0

# now `model.forward` is what we call here an "instance method", and now we want to patch it

def our_forward(x):
    return 42

patcher = Patcher()
patcher.patch(model.forward, our_forward)

assert model.forward(7) == 42 
```

Try implementing that test case, maybe debugging the flow of the patcher for this case to see if it works as expected (or not). If  you have further questions, feel free to post these. I think @nikita-savelyevv might also pitch in with some help on these.

**Comment by nikita-savelyevv** [comment]
Hi @Candyzorua. Regarding your question, if you call `patch()` on a method of the instantiated class object then it should fall into the branch you mention. This for example will happen if you call
```
PATCHER.patch(test_obj.assert_wrapper_stack_method, wrapper1)
```

for the `test_obj` which is created at https://github.com/openvinotoolkit/nncf/blob/develop/tests/common/test_monkey_patching.py#L51 .

**Comment by Candyzorua** [comment]
@vshampor @nikita-savelyevv  Thank you both your input. Duly noted and I will implement the ideas you have mentioned at my earliest availability.

**Comment by MaximProshin** [comment]
> Hello, I wrote unit tests for the first 5 files and have made the following [PR](https://github.com/openvinotoolkit/nncf/pull/2468), am continuing with the rest of the tests :)

https://github.com/openvinotoolkit/nncf/pull/2468 has been merged

**Comment by Candyzorua** [comment]
Hello again,

I have been writing tests for nncf/quantization recently and encountered an issue while making tests for [nncf/quantization/algorithms/accuracy_control/algorithm.py](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/algorithm.py). 

The function I am testing is `_collect_original_biases_and_weights`.  The test code is as such:

<img width="342" alt="image" src="https://github.com/openvinotoolkit/openvino/assets/75513525/5959aaa6-8e7b-4ca5-a5c2-c16090c2238c">

I get a synthetic linear model from [nncf/tests/openvino/native/models.py](https://github.com/openvinotoolkit/nncf/blob/develop/tests/openvino/native/models.py) and quantize it. However, when I pass this to `_collect_original_biases_and_weights` I get an error because the nodes of the quantized graph's have None for layer attributes. 

Not sure if the  is a bug, or if I have done something incorrect in the testing setup. Any help would be appreciated!

**Comment by Candyzorua** [comment]
In addition, I would like to clarify if the method through which I am obtaining synthetic models and quantizing them is good practice, or if I should hardcode quantized models.

**Comment by vshampor** [comment]
@andrey-churkin to advise on the accuracy control algorithm requirements
